lib/SILOptimizer/SILCombiner/SILCombinerCastVisitors.cpp

```diff
@@ -282,35 +387,92 @@ SILCombiner::visitUncheckedAddrCastInst(UncheckedAddrCastInst *UADCI) {
-    return Builder.createUpcast(URCI->getLoc(), URCI->getOperand(),
-                                URCI->getType());
-
-  // (unchecked_ref_cast (open_existential_ref (init_existential_ref X))) ->
-  // (unchecked_ref_cast X)
-  if (auto *OER = dyn_cast<OpenExistentialRefInst>(URCI->getOperand()))
-    if (auto *IER = dyn_cast<InitExistentialRefInst>(OER->getOperand()))
-      return Builder.createUncheckedRefCast(URCI->getLoc(), IER->getOperand(),
-                                            URCI->getType());
+  // %0 = upcast %x : $X->Y
+  // %1 = unchecked_ref_cast %0 : $Y->Z
+  //
+  // ->
+  //
+  // %0 = upcast %x : $X->Y
+  // %1 = unchecked_ref_cast %x : $X->Z
+  //
+  // NOTE: For owned values, we only perform this optimization if we can
+  // guarantee that we can eliminate the upcast.
+  if (auto *ui = dyn_cast<UpcastInst>(urci->getOperand())) {
+    SILValue uiOp = ui->getOperand();
+
+    if (uiOp.getOwnershipKind() != OwnershipKind::Owned) {
+      return Builder.createUncheckedRefCast(urci->getLoc(), uiOp,
+                                            urci->getType());
+    }
+
+    SingleBlockOwnedForwardingInstFolder folder(*this, urci);
+    if (folder.add(ui)) {
+      auto *newValue =
+          Builder.createUncheckedRefCast(urci->getLoc(), uiOp, urci->getType());
+      return std::move(folder).optimizeWithReplacement(newValue);
+    }
+  }
+
+  // This is an exact transform where we are replacing urci with an upcast on
+  // the same value. So from an ownership perspective because both instructions
+  // are forwarding and we are eliminating urci, we are safe.
+  if (urci->getType() != urci->getOperand()->getType() &&
+      urci->getType().isExactSuperclassOf(urci->getOperand()->getType()))
+    return Builder.createUpcast(urci->getLoc(), urci->getOperand(),
+                                urci->getType());
+
+  // %0 = init_existential_ref %x : $X -> Existential
+  // %1 = open_existential_ref %0 : $Existential -> @opened() Existential
+  // %2 = unchecked_ref_cast %1
+  //
+  // ->
+  //
+  // %0 = init_existential_ref %x : $X -> Existential
+  // %1 = open_existential_ref %0 : $Existential -> @opened() Existential
+  // %2 = unchecked_ref_cast %x
+  //
+  // NOTE: When we have an owned value, we only perform this optimization if we
+  // can remove both the open_existential_ref and the init_existential_ref.
+  if (auto *oer = dyn_cast<OpenExistentialRefInst>(urci->getOperand())) {
+    if (auto *ier = dyn_cast<InitExistentialRefInst>(oer->getOperand())) {
+      if (ier->getOwnershipKind() != OwnershipKind::Owned) {
+        return Builder.createUncheckedRefCast(urci->getLoc(), ier->getOperand(),
+                                              urci->getType());
+      }
+
+      SingleBlockOwnedForwardingInstFolder folder(*this, urci);
+      if (folder.add(oer) && folder.add(ier)) {
+        auto *newValue = Builder.createUncheckedRefCast(
+            urci->getLoc(), ier->getOperand(), urci->getType());
```

Comment:
Still not sure why you create a new instruction

Reply:
Also matching what the code did previously. It wasn't clear to me, why this needed to be done, so I just matched it.